File: ruby/BinaryTree.rb
class BinaryTree
  attr_reader :value, :right, :left
  attr_writer :right, :left

  def initialize(value)
    @value = value
    @left  = nil
    @right = nil
  end

end

def breadthFirstLevel(tree)
  q = Array.new
  q << tree
  print = Array.new
  level = 0
  currentCount = 1
  c = 0
  while q.length > 0
    t = q.shift
    currentCount -= 1
    print t.value + " "

    if not t.left.nil?
      q << t.left
      c += 1
    end

    if not t.right.nil?
      q << t.right
      c += 1
    end

    if currentCount == 0
      print "\n"
      currentCount = c
      c = 0
    end
  end
end

def depthFirstPreOrder(tree)
  print tree.value + " "
  if not tree.left.nil?
    depthFirstPreOrder(tree.left)
  end
  if not tree.right.nil?
    depthFirstPreOrder(tree.right)
  end
end

def depthFirstInOrder(tree)
  if not tree.left.nil?
    depthFirstInOrder(tree.left)
  end
  print tree.value + " "
  if not tree.right.nil?
    depthFirstInOrder(tree.right)
  end
end

def depthFirstPostOrder(tree)
  if not tree.left.nil?
    depthFirstPostOrder(tree.left)
  end
  if not tree.right.nil?
    depthFirstPostOrder(tree.right)
  end
  print tree.value + " "
end

File: scala/BinaryTree.scala
import scala.collection.mutable.Queue

class BinaryTreeNode(v: String) {
  var value: String = v;
  var left:  BinaryTreeNode = null;
  var right: BinaryTreeNode = null;
}

class BinaryTree(h: BinaryTreeNode) {
  var head: BinaryTreeNode = h;

  def breadthFirstLevelOrderPrint(tree: BinaryTreeNode = head) {
    var q = new Queue[BinaryTreeNode];
    q.enqueue(tree);
    var currentCount: Int = 1;
    var c: Int = 0;
    while (q.length > 0) {
      var node: BinaryTreeNode = q.dequeue;
      currentCount -= 1;
      print(node.value + " ");

      if (node.left != null) {
        q.enqueue(node.left);
        c += 1;
      }
      if (node.right != null) {
        q.enqueue(node.right);
        c += 1;
      }

      if (currentCount == 0) {
        println;
        currentCount = c;
        c = 0;
      }
    }
  }

  def depthFirstPreOrderPrint(tree: BinaryTreeNode = head) {
    print(tree.value + " ");
    if (tree.left != null) {
      depthFirstPreOrderPrint(tree.left);
    }
    if (tree.right != null) {
      depthFirstPreOrderPrint(tree.right);
    }
  }

  def depthFirstInOrderPrint(tree: BinaryTreeNode = head) {
    if (tree.left != null) {
      depthFirstInOrderPrint(tree.left);
    }
    print(tree.value + " ");
    if (tree.right != null) {
      depthFirstInOrderPrint(tree.right);
    }
  }

  def depthFirstPostOrderPrint(tree: BinaryTreeNode = head) {
    if (tree.left != null) {
      depthFirstPostOrderPrint(tree.left);
    }
    if (tree.right != null) {
      depthFirstPostOrderPrint(tree.right);
    }
    print(tree.value + " ");
  }
}
